### PR TITLE
Laboratory work #2, Tasya Kakoulina - 23FPL2

### PR DESCRIPTION
**Pull request by tasya-kakoulina**

### CODE REVIEW

File: lab_2_retrieval_w_bm25/main.py [hunk 1]
```diff
@@ -18,6 +20,18 @@ def tokenize(text: str) -> list[str] | None:
...
+        return None
+    result = []
+    word = ''
+    for token in text.lower():
+        if not token.isalpha():
+            if word:
+                result.append(word)
+            word = ''
+            continue
+        word += token
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:33]: join is better

File: lab_2_retrieval_w_bm25/main.py [hunk 2]
```diff
@@ -33,6 +47,15 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
     """
+    if not isinstance(tokens, list) or not all(isinstance(token, str) for token in tokens):
+        return None
+    if (not stopwords or not isinstance(stopwords, list)
+            or not all(isinstance(stopword, str) for stopword in stopwords)):
+        return None
+    for i in tokens[:]:
+        if i in stopwords:
+            tokens.remove(i)
+    return tokens if tokens else None
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:58]: `tokens or None`

File: lab_2_retrieval_w_bm25/main.py [hunk 3]
```diff
@@ -120,6 +194,31 @@ def calculate_bm25(
...
+    bm_dict = {}
+    unique_words = list(set(vocab + document))
+    for token in unique_words:
+        numerator = document.count(token) * (k1 + 1)
+        denominator = document.count(token) + k1 * (1 - b + b * doc_len/avg_doc_len)
+        if token not in idf_document:
+            bm_dict[token] = 0.0
+        else:
+            bm_dict[token] = idf_document[token] * numerator / denominator
+    return bm_dict if bm_dict else None
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:221]: .

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -138,6 +237,26 @@ def rank_documents(
...
+        return None
+    res = []
+    for ind, dictionary in enumerate(indexes):
+        num = 0.0
+        for word in query_words:
+            if word in dictionary:
+                num += dictionary[word]
+        res.append((ind, num))
+    res.sort(key=lambda x: x[-1], reverse=True)
+    return res if res else None
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:259]: .

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Perfect

File: lab_2_retrieval_w_bm25/main.py (hunk 1 above)
Comment (tasya-kakoulina) [reply lab_2_retrieval_w_bm25/main.py]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 2 above)
Comment (tasya-kakoulina) [reply lab_2_retrieval_w_bm25/main.py]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 3 above)
Comment (tasya-kakoulina) [reply lab_2_retrieval_w_bm25/main.py]: done

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_retrieval_w_bm25/main.py [hunk 5]
```diff
@@ -33,6 +47,12 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(tokens, list) or not all(isinstance(token, str) for token in tokens):
+        return None
+    if (not stopwords or not isinstance(stopwords, list) or
+            not all(isinstance(stopword, str) for stopword in stopwords)):
+        return None
+    return [token for token in tokens if token not in stopwords] or None
```
[the remark this answers is not in the document]
Comment (tasya-kakoulina) [reply lab_2_retrieval_w_bm25/main.py:55]: didn't get it :( should i use sorted here? or in rank_documents? but why?..

File: lab_2_retrieval_w_bm25/main.py [hunk 6]
```diff
@@ -47,6 +67,16 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
...
     """
+    if not isinstance(documents, list):
+        return None
+    for document in documents:
+        if not isinstance(document, list):
+            return None
+        for word in document:
+            if not isinstance(word, str):
+                return None
+    vocab = set(sum(documents, []))
```
[the remark this answers is not in the document]
Comment (tasya-kakoulina) [reply lab_2_retrieval_w_bm25/main.py:78]: me too!

**Pull request closed without merging**